lib/msf/util/windows_registry_parser.rb

```diff
@@ -0,0 +1,362 @@
+        end
+        return root_key if root_key
+      rescue IOError
+        raise StandardError, 'Cannot parse the RegHbin structure'
+      end
+      raise StandardError, 'Cannot find the RootKey' unless reg_hbin
+    end
+
+    def get_value(reg_key, reg_value)
+      reg_key = find_key(reg_key)
+      return nil unless reg_key
+
+      if reg_key.data.num_values > 0
+        value_list = get_value_blocks(reg_key.data.offset_value_list, reg_key.data.num_values + 1)
+        value_list.each do |value|
+          if value.data.name == reg_value ||
+             reg_value.downcase == 'default' && value.data.flag <= 0
+            return value.data.value_type, get_value_data(value.data)
+          end
+        end
+      end
+      nil
+    end
+
+    def find_key(key)
+      # Let's strip '\' from the beginning, except for the case of
+      # only asking for the root node
+      key = key[1..-1] if key[0] == '\\' && key.size > 1
+
+      parent_key = @root_key
+      if key.size > 0 && key[0] != '\\'
+        key.split('\\').each do |sub_key|
+          res = find_sub_key(parent_key, sub_key)
+          return nil unless res
+          parent_key = res
+        end
+      end
+      parent_key
+    end
+
+    def find_sub_key(parent_key, sub_key)
+      unless parent_key&.data&.magic == NK_MAGIC
+        raise ArgumentError, "find_sub_key: parent key must be a NK record"
+      end
+      block = get_block(parent_key.data.offset_sub_key_lf)
+      blocks = []
+      if block.data.magic == RI_MAGIC
+        # ri points to lf/lh records, so we consolidate them in the main blocks array
+        block.data.hash_records.each do |hash_record|
+          blocks << get_block(hash_record.offset_nk)
+        end
+      else
+        blocks << block
+      end
+
+      # Let's search the hash records for the name
+      blocks.each do |block|
+        block.data.hash_records.each do |hash_record|
+          res = compare_hash(block.data.magic, hash_record, sub_key)
+          if res
+            nk = get_block(res)
+            return nk if nk.data.key_name == sub_key
+          end
+        end
+      end
+
+      nil
+    end
+
+    def get_block(offset)
+      RegHbinBlock.read(@hive_data[4096+offset..-1])
+    end
+
+    def compare_hash(magic, hash_rec, key)
+      case magic
+      when LF_MAGIC
+        if hash_rec.key_name.gsub(/(^\x00*)|(\x00*$)/, '') == key[0,4]
+          return hash_rec.offset_nk
+        end
+      when LH_MAGIC
+        if hash_rec.key_name.unpack('<L').first == get_lh_hash(key)
+          return hash_rec.offset_nk
+        end
+      when RI_MAGIC
+        # Special case here, don't know exactly why, an RI pointing to a NK
+        offset = hash_rec.offset_nk
+        nk = get_block(offset)
+        return offset if nk.key_name == key
+      else
+        raise ArgumentError, "Unknow magic: #{magic}"
+      end
+    end
+
+    def get_lh_hash(key)
+      res = 0
+      key.upcase.bytes do |byte|
+        res *= 37
```

Comment:
What is 37 here?

Reply:
This is from the 'lh' subkey-list hash algorithm (see [Appendix C](http://www.sentinelchicken.com/data/TheWindowsNTRegistryFileFormat.pdf)). This was taken from `impacket` as is, but I will add the reference in a comment.